### PR TITLE
Fix backward compatability for older versions of lightning

### LINKED ISSUE #152: Retrieval metrics problem with pytorch lightning integration in compute()
## 🐛 Bug

I use the commit https://github.com/PyTorchLightning/metrics/commit/f06488faf79d4f4792cd392e964870d4898dde45 to calculate `RetrievalMAP` and `RetrievalPrecision` in a pytorch-lightning module. The `validation_step` and `validation_step_end` functions work, but the `fast_dev_run=True` gives an error in the `compute()` step.

However, when I run `self.log("val_MAP", self.metric.compute())` instead of `self.log("val_MAP", self.metric)` in the `validation_step_end` I do not get errors. But computing the whole metric is becoming very slow if it is done every validation_step.

### To Reproduce

I run the following code with the mentioned commit. 

#### Code sample

```python
from typing import Optional
import os
import torch
from torch import nn
import torch.nn.functional as F
from torchvision import transforms
from torchvision.datasets import MNIST
from torch.utils.data import DataLoader, random_split
import pytorch_lightning as pl

from torchmetrics import (
    RetrievalMAP,
    RetrievalPrecision,
    MeanAbsoluteError,
)


class MNISTDataModule(pl.LightningDataModule):
    def __init__(self, batch_size=32):
        super().__init__()
        self.batch_size = batch_size

    def prepare_data(self):
        MNIST(os.getcwd(), train=True, download=True)
        MNIST(os.getcwd(), train=False, download=True)

    def setup(self, stage: Optional[str] = None):
        transform = transforms.Compose(
            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
        )
        if stage == "fit":
            mnist_train = MNIST(os.getcwd(), train=True, transform=transform)
            self.mnist_train, self.mnist_val = random_split(
                mnist_train, [55000, 5000]
            )
        if stage == "test":
            self.mnist_test = MNIST(
                os.getcwd(), train=False, transform=transform
            )

    def train_dataloader(self):
        mnist_train = DataLoader(self.mnist_train, batch_size=self.batch_size)
        return mnist_train

    def val_dataloader(self):
        mnist_val = DataLoader(self.mnist_val, batch_size=self.batch_size)
        return mnist_val

    def test_dataloader(self):
        mnist_test = DataLoader(self.mnist_test, batch_size=self.batch_size)
        return mnist_test


class LitAutoEncoder(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 1), nn.Softplus()
        )
        self.decoder = nn.Sequential(
            nn.Linear(1, 64), nn.ReLU(), nn.Linear(64, 28 * 28)
        )
        # self.metric = RetrievalMAP()
        self.metric = RetrievalPrecision()
        # self.metric = MeanAbsoluteError()

    def forward(self, x):
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        x = x.view(x.size(0), -1)
        preds = self.encoder(x).squeeze()

        indexes = torch.randint(100, size=preds.size())
        targets = torch.randint(2, size=preds.size()).to(bool)

        return {"indexes": indexes, "preds": preds, "targets": targets}

    def validation_step_end(self, outputs):
        self.metric(outputs["indexes"], outputs["preds"], outputs["targets"])
        # self.metric(outputs["preds"], outputs["preds"] ** 2)
        self.log("val_MAP", self.metric)
        # self.log("val_MAP", self.metric.compute())

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


if __name__ == "__main__":
    datamodule = MNISTDataModule()
    module = LitAutoEncoder()

    trainer = pl.Trainer(gpus=1, fast_dev_run=True)

    trainer.fit(module, datamodule=datamodule)
    trainer.test(module, datamodule=datamodule)
```

#### StackTrace

```
Traceback (most recent call last):
  File "reproduce_retrieval_error.py", line 107, in <module>
    trainer.fit(module, datamodule=datamodule)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 114, in start_training
    self._results = trainer.run_train()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 637, in run_train
    self.train_loop.run_training_epoch()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/training_loop.py", line 577, in run_training_epoch
    self.trainer.run_evaluation(on_epoch=True)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 754, in run_evaluation
    eval_loop_results = self.evaluation_loop.log_epoch_metrics_on_evaluation_end()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 200, in log_epoch_metrics_on_evaluation_end
    eval_loop_results = self.trainer.logger_connector.get_evaluate_epoch_results()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py", line 286, in get_evaluate_epoch_results
    metrics_to_log = self.cached_results.get_epoch_log_metrics()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 405, in get_epoch_log_metrics
    return self.run_epoch_by_func_name("get_epoch_log_metrics")
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 398, in run_epoch_by_func_name
    results = [func() for func in results]
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 398, in <listcomp>
    results = [func() for func in results]
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 128, in get_epoch_log_metrics
    return self.get_epoch_from_func_name("get_epoch_log_metrics")
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 121, in get_epoch_from_func_name
    self.run_epoch_func(results, opt_metrics, func_name, *args, **kwargs)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py", line 110, in run_epoch_func
    metrics_to_log = func(*args, add_dataloader_idx=self.has_several_dataloaders, **kwargs)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/pytorch_lightning/core/step_result.py", line 327, in get_epoch_log_metrics
    result[dl_key] = self[k].compute().detach()
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/torchmetrics/metric.py", line 228, in wrapped_func
    self._computed = compute(*args, **kwargs)
  File "/home/netter/.cache/pypoetry/virtualenvs/stereographic-link-prediction-ra14Y8Aq-py3.8/lib/python3.8/site-packages/torchmetrics/retrieval/retrieval_metric.py", line 110, in compute
    idx = torch.cat(self.idx, dim=0)
RuntimeError: There were no tensor arguments to this function (e.g., you passed an empty list of Tensors), but no fallback function is registered for schema aten::_cat.  This usually means that this function requires a non-empty list of Tensors.  Available functions are [CPU, CUDA, QuantizedCPU, BackendSelect, Named, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradNestedTensor, UNKNOWN_TENSOR_TYPE_ID, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, Autocast, Batched, VmapMode].

CPU: registered at /pytorch/build/aten/src/ATen/RegisterCPU.cpp:5925 [kernel]
CUDA: registered at /pytorch/build/aten/src/ATen/RegisterCUDA.cpp:7100 [kernel]
QuantizedCPU: registered at /pytorch/build/aten/src/ATen/RegisterQuantizedCPU.cpp:641 [kernel]
BackendSelect: fallthrough registered at /pytorch/aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]
Named: registered at /pytorch/aten/src/ATen/core/NamedRegistrations.cpp:7 [backend fallback]
AutogradOther: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradCPU: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradCUDA: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradXLA: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradNestedTensor: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
UNKNOWN_TENSOR_TYPE_ID: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradPrivateUse1: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradPrivateUse2: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
AutogradPrivateUse3: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9122 [autograd kernel]
Tracer: registered at /pytorch/torch/csrc/autograd/generated/TraceType_2.cpp:10525 [kernel]
Autocast: registered at /pytorch/aten/src/ATen/autocast_mode.cpp:254 [kernel]
Batched: registered at /pytorch/aten/src/ATen/BatchingRegistrations.cpp:1016 [backend fallback]
VmapMode: fallthrough registered at /pytorch/aten/src/ATen/VmapModeRegistrations.cpp:33 [backend fallback]
```

### Expected behavior

This error should not show up. I expect the metric to be computed correctly. When I use the `MeanAbsoluteError` as metric the code works. Therefore, there must be a bug in the `compute` step of the Retrieval Metrics in combination with pytorch-lightnings API as a call to `compute()` within the `validation_step_end` does not create errors.

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1+cu102
 - OS (e.g., Linux): Ubuntu on WSL2
 - How you installed PyTorch (`conda`, `pip`, source): pip / poetry
 - Build command you used (if compiling from source):
 - Python version: 3.8.8
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 1 GPU
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #152

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
can we also add an integration test for this as we have tested with PL 1.2 and they were all passing...

File: torchmetrics/metric.py [hunk 1]
```diff
@@ -256,7 +257,10 @@ def reset(self):
         """
         This method automatically resets the metric state variables to their default value.
         """
-        self._computed = None
+        # lower lightning versions requires this implicitly to log metric objects correctly
+        # in self.log
+        if _LIGHTNING_GREATER_THAN_1_2_8:
+            self._computed = None
```
Comment (Borda) [inline torchmetrics/metric.py:263]: was this already fixed in PL 1.2.8?
at least my reading of this that there is a change in 1.2.8 and from this version will be used...

**Comment by Borda** [comment]
The problem seems to be cyclic import as the available func tries to import the package but in PL case it has TM as a dependency...

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@SkafteNicki seems that the integration fails... mind check it?

**Review by justusschock — APPROVED** [review]

File: torchmetrics/metric.py [hunk 2]
```diff
@@ -256,7 +257,10 @@ def reset(self):
         """
         This method automatically resets the metric state variables to their default value.
         """
-        self._computed = None
+        # lower lightning versions requires this implicitly to log metric objects correctly
+        # in self.log
+        if not _LIGHTNING_AVAILABLE or _LIGHTNING_GREATER_EQUAL_1_3:
+            self._computed = None
```
Comment (ananthsub) [inline torchmetrics/metric.py:263]: isn't this backwards? shouldn't lightning be accounting for the torchmetrics version, not the other way around?

Comment (SkafteNicki) [reply torchmetrics/metric.py:263]: since we cannot change earlier versions of lightning, we need to do the change in torchmetrics

**Pull request merged**